### PR TITLE
Add HARD token governance committee for Hard module

### PR DESCRIPTION
**Pull request by denalimarsh**
Creates a HARD token governance committee responsible for managing Hard module's parameters.

Question: do we want the stability committee to maintain its current access to the same Hard module parameters?

### CODE REVIEW

File: migrate/v0_15/migrate.go [hunk 1]
```diff
@@ -174,11 +175,49 @@ func Committee(genesisState v0_14committee.GenesisState) v0_15committee.GenesisS
 			baseStabilityCom := stabilityCom.SetPermissions(newStabilityCommitteePermissions)
 			newStabilityCom := v0_15committee.MemberCommittee{BaseCommittee: baseStabilityCom}
 			committees = append(committees, newStabilityCom)
-		} else {
+		case 2:
 			safetyCom := comtypes.NewMemberCommittee(com.ID, com.Description, com.Members,
 				[]v0_15committee.Permission{v0_15committee.SoftwareUpgradePermission{}},
 				com.VoteThreshold, com.ProposalDuration, v0_15committee.FirstPastThePost)
 			committees = append(committees, safetyCom)
+		case 3:
```
Comment (karzak) [inline migrate/v0_15/migrate.go:183]: See comments on #940

**Review by karzak — CHANGES_REQUESTED** [review]

File: migrate/v0_15/migrate.go (hunk 1 above)
Comment (denalimarsh) [reply migrate/v0_15/migrate.go]: Updated in latest

**Comment by denalimarsh** [comment]
@karzak do we want the stability committee to maintain its current access to the Hard module parameters?

**Review by karzak — CHANGES_REQUESTED** [review]
Was playing around with exporting the resulting state to JSON and got a panic - need to register `BaseCommittee` in codec.go

```
cdc.RegisterConcrete(BaseCommittee{}, "kava/BaseCommittee", nil)
```

One other nit, otherwise LGTM

**Comment by karzak** [comment]
> @karzak do we want the stability committee to maintain its current access to the Hard module parameters?

Good question, will have to think about it more.  Added issue to to-do.

File: migrate/v0_15/migrate.go [hunk 2]
```diff
@@ -174,14 +175,50 @@ func Committee(genesisState v0_14committee.GenesisState) v0_15committee.GenesisS
...
 				com.VoteThreshold, com.ProposalDuration, v0_15committee.FirstPastThePost)
 			committees = append(committees, safetyCom)
 		}
 	}
 
+	// Initialize hard governance committee
+	hardGovCom := comtypes.TokenCommittee{}
+	hardGovCom.ID = 3
+	hardGovCom.Description = "Hard Governance Committee"
+	hardGovCom.Members = []sdk.AccAddress{sdk.AccAddress("kava1e0agyg6eug9r62fly9sls77ycjgw8ax6xk73es")}
```
Comment (karzak) [inline migrate/v0_15/migrate.go:190]: what address is this? It also doesn't produce the expected address, members shows up as

```
kava1ddshvcf3v5cxzemevumx2at889ervvnxd3unjumvwvmnw7trdfnhwwrp0qm8s6ehxdjhxaff7jp
```

File: migrate/v0_15/migrate.go [hunk 3]
```diff
@@ -174,14 +175,50 @@ func Committee(genesisState v0_14committee.GenesisState) v0_15committee.GenesisS
...
+	hardGovCom.VoteThreshold = sdk.MustNewDecFromStr("0.5")
+	hardGovCom.ProposalDuration = time.Duration(604800000000000)
+	hardGovCom.TallyOption = v0_15committee.FirstPastThePost
```
Comment (karzak) [inline migrate/v0_15/migrate.go:193]: token holder committees should be tallied at deadline

File: migrate/v0_15/migrate.go [hunk 4]
```diff
@@ -174,14 +175,50 @@ func Committee(genesisState v0_14committee.GenesisState) v0_15committee.GenesisS
...
 		}
 	}
 
+	// Initialize hard governance committee
+	hardGovCom := comtypes.TokenCommittee{}
+	hardGovCom.ID = 3
+	hardGovCom.Description = "Hard Governance Committee"
+	hardGovCom.Members = []sdk.AccAddress{sdk.AccAddress("kava1e0agyg6eug9r62fly9sls77ycjgw8ax6xk73es")}
+	hardGovCom.VoteThreshold = sdk.MustNewDecFromStr("0.5")
+	hardGovCom.ProposalDuration = time.Duration(604800000000000)
```
Comment (karzak) [inline migrate/v0_15/migrate.go:192]: nit: explicitly define duration using predefined units (time.Hour, etc) - too easy to make a mistake manually adding zeros

File: migrate/v0_15/migrate.go [hunk 5]
```diff
@@ -174,14 +175,50 @@ func Committee(genesisState v0_14committee.GenesisState) v0_15committee.GenesisS
...
-		} else {
+		case 2:
 			safetyCom := comtypes.NewMemberCommittee(com.ID, com.Description, com.Members,
 				[]v0_15committee.Permission{v0_15committee.SoftwareUpgradePermission{}},
 				com.VoteThreshold, com.ProposalDuration, v0_15committee.FirstPastThePost)
 			committees = append(committees, safetyCom)
 		}
 	}
 
+	// Initialize hard governance committee
```
Comment (karzak) [inline migrate/v0_15/migrate.go:186]: I would initialize using the 'NewTokenCommittee' method

File: migrate/v0_15/migrate.go [hunk 6]
```diff
@@ -174,14 +175,50 @@ func Committee(genesisState v0_14committee.GenesisState) v0_15committee.GenesisS
...
+	hardGovCom := comtypes.TokenCommittee{}
+	hardGovCom.ID = 3
+	hardGovCom.Description = "Hard Governance Committee"
+	hardGovCom.Members = []sdk.AccAddress{sdk.AccAddress("kava1e0agyg6eug9r62fly9sls77ycjgw8ax6xk73es")}
+	hardGovCom.VoteThreshold = sdk.MustNewDecFromStr("0.5")
+	hardGovCom.ProposalDuration = time.Duration(604800000000000)
+	hardGovCom.TallyOption = v0_15committee.FirstPastThePost
+	hardGovCom.Quorum = sdk.MustNewDecFromStr("0.33")
+	hardGovCom.TallyDenom = "hard"
```
Comment (karzak) [inline migrate/v0_15/migrate.go:195]: `Quorum` and `TallyDenom` are not getting preserved when marshaling to JSON.

File: migrate/v0_15/migrate_test.go [hunk 7]
```diff
@@ -41,7 +41,7 @@ func TestCommittee(t *testing.T) {
 	err = newGenState.Validate()
 	require.NoError(t, err)
 
-	require.Equal(t, len(oldGenState.Committees), len(newGenState.Committees))
+	require.Equal(t, len(oldGenState.Committees), len(newGenState.Committees)-1) // New gen state has 1 more committee
```
Comment (karzak) [inline migrate/v0_15/migrate_test.go:44]: At the end of the test, export the resulting state 

```
v15Cdc := app.MakeCodec()
	ioutil.WriteFile(filepath.Join("testdata", "kava-8-committee-state.json"), v15Cdc.MustMarshalJSON(newGenState), 0644)
```

There's multiple errors in the way the committees are exported

1. Each committee has a top-level `Type` field, as well as `BaseCommittee.Type`. Is this correct? Why the redundant field?
2. The hard committee top-level type field is not populated.
3. Hard committee is missing quorum and tally_denom

**Review by karzak — CHANGES_REQUESTED** [review]

File: migrate/v0_15/migrate.go [hunk 8]
```diff
@@ -174,14 +175,50 @@ func Committee(genesisState v0_14committee.GenesisState) v0_15committee.GenesisS
...
+		v0_15committee.AllowedParam{Subspace: "hard", Key: "MoneyMarkets"},
+		v0_15committee.AllowedParam{Subspace: "hard", Key: "MinimumBorrowUSDValue"},
+	}
+	newHardSubParamPermissions.AllowedParams = newAllowedParams
+
+	// Money market permissions
+	var newMoneyMarketParams v0_15committee.AllowedMoneyMarkets
+	hardMMDenoms := []string{"bnb", "busd", "btcb", "xrpb", "usdx", "ukava", "hard"}
+	for _, mmDenom := range hardMMDenoms {
+		newMoneyMarketParam := v0_15committee.NewAllowedMoneyMarket(mmDenom, true, false, false, true, true, true)
```
[the remark this answers is not in the document]
Comment (denalimarsh) [reply migrate/v0_15/migrate.go:212]: If I'm understanding correctly, this should already be possible. The committee has permission to change the `hard.MoneyMarkets` param, where each `MoneyMarket` contains the `SpotMarketID`.

File: migrate/v0_15/migrate.go (hunk 5 above)
Comment (denalimarsh) [reply migrate/v0_15/migrate.go:194]: Updated in https://github.com/Kava-Labs/kava/pull/941/commits/25ea9cb6caa3db758836a1f8191ccf586aacae5f

File: migrate/v0_15/migrate.go (hunk 4 above)
Comment (denalimarsh) [reply migrate/v0_15/migrate.go]: Addressed in https://github.com/Kava-Labs/kava/pull/941/commits/25ea9cb6caa3db758836a1f8191ccf586aacae5f

File: migrate/v0_15/migrate.go (hunk 3 above)
Comment (denalimarsh) [reply migrate/v0_15/migrate.go]: Updated in https://github.com/Kava-Labs/kava/pull/941/commits/25ea9cb6caa3db758836a1f8191ccf586aacae5f

File: migrate/v0_15/migrate.go (hunk 2 above)
Comment (denalimarsh) [reply migrate/v0_15/migrate.go]: Addressed in https://github.com/Kava-Labs/kava/pull/941/commits/25ea9cb6caa3db758836a1f8191ccf586aacae5f

File: migrate/v0_15/migrate.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (denalimarsh) [reply migrate/v0_15/migrate.go]: Updated in https://github.com/Kava-Labs/kava/pull/941/commits/25ea9cb6caa3db758836a1f8191ccf586aacae5f

File: migrate/v0_15/migrate.go (hunk 6 above)
Comment (denalimarsh) [reply migrate/v0_15/migrate.go]: The issue was that the `Committee` interface's `SetPermissions` function always returned `BaseCommittee`, which caused the extra fields to be dropped from TokenCommittee in the migration after permissions were set. Fixed in https://github.com/Kava-Labs/kava/pull/941/commits/7ef2087dde9d915a1b79ddcbf6d95c1d4802e041 by updating the return type to the `Committee` interface and adding `SetPermissions` methods to each sub-committee type.

File: migrate/v0_15/migrate_test.go (hunk 7 above)
Comment (denalimarsh) [reply migrate/v0_15/migrate_test.go:44]: 1. Fixed in https://github.com/Kava-Labs/kava/pull/941/commits/d930aef7c30fefe8234d9bdb6f359ca60b01bdc1 by removing `BaseCommittee`'s Type field and updating the `GetType` methods. Now there's only one top-level `Type` field.
2. Fixed in https://github.com/Kava-Labs/kava/commit/7ef2087dde9d915a1b79ddcbf6d95c1d4802e041
3. Fixed in https://github.com/Kava-Labs/kava/commit/7ef2087dde9d915a1b79ddcbf6d95c1d4802e041

**Comment by denalimarsh** [comment]
Whew, lots of revisions on this one. Thanks for the tip about exporting state for manual review at the end of the test, it was helpful for finding issues in the clunky JSON marshaler. Things should be all patched up now 👍

**Review by karzak — APPROVED** [review]
tACK - revisions look good, tested manually with kvtool and working as expected.

**Pull request merged**